components/events/src/event_bus.cpp

```diff
@@ -0,0 +1,73 @@
+/* Copyright (c) 2022, Arm Limited and Contributors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 the "License";
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "components/events/event_bus.hpp"
+
+#include <cassert>
+
+namespace components
+{
+namespace events
+{
+EventBus &EventBus::attach(std::weak_ptr<EventObserver> &&observer)
+{
+	for (auto it = m_observers.begin(); it != m_observers.end(); it++)
```

Comment:
Is adding the same observer multiple times supposed to be valid? Or should you just assert that it's not yet attached?

Reply:
Adding the same observer object multiple times is technically valid behavior as we can just ignore the addition. However, I have added an assertion as this may indicate something else is wrong in the event bus usage.

Reply:
Maybe a better assert, without explicit loop:
`assert((std::find_if(m_observers.begin(), m_observers.end(), [observer](auto const & obs){return same_ptr(obs, observer);}) == m_observers.end()) && "attempting...");`